pallets/randomness/src/types.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{traits::*, BalanceOf, Config, Error, Pallet};
+use frame_support::pallet_prelude::*;
+use frame_support::traits::{Currency, ExistenceRequirement::KeepAlive, ReservableCurrency};
+use frame_support::weights::WeightToFeePolynomial;
+use sp_runtime::traits::{CheckedSub, Saturating};
+
+#[derive(PartialEq, Copy, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+/// Randomness storage item from BABE
+pub enum BabeRandomness {
+	OneEpochAgo,
+	TwoEpochsAgo,
+	CurrentBlock,
+}
+
+#[derive(PartialEq, Copy, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+/// Information regarding the request type
+pub enum RequestType<BlockNumber> {
+	/// BABE randomness from reading relay chain state proof
+	Babe {
+		info: BabeRandomness,
+		when: BlockNumber,
```

Comment:
```suggestion
		when: enum { BlockNumber, EpochIndex },
```
add path to allow requesting BABE randomness by epoch instead of by block number

before wasn't considered possible because we didn't get the epoch index but now we will have it